### DISCUSSION TITLE
forgejo support

### DISCUSSION THREAD

**Discussion by adamcstephens** [cloudbase/garm#303] [Ideas]
Have you considered supporting other forges, such as [forgejo](https://forgejo.org/)? They have a similar runner setup based on github's, and could allow for fully self-hosted environments.

Codeberg is running Forgejo and you can see their docs here https://docs.codeberg.org/ci/actions/

**Comment by gabriel-samfira** [comment]
Hi @adamcstephens 

It depends. The runner software is just one part of the whole thing. We also rely on webhooks, the ability to set labels on the runner, ephemeral runners, etc.

There are several layers that need to exist to successfully work in the current code paths. We can look into it as time permits, but no promises.

**Comment by adamcstephens** [comment]
Thanks! I understand there's no promises but wanted to potentially get it on your radar and hear thoughts on feasibility. I realize there's a lot of pieces to it. :)

**Reply by gabriel-samfira** [reply]
So we now have a Gitea branch here:
  * https://github.com/cloudbase/garm/pull/393

I realize it's not the same as forgejo, but Gitea so far has everything we need to "easily" add support for it. By easily I mean it has the same functionality that allows us to not change many assumptions we make for github (workflow_job webhook, ephemeral runners that get removed once a job finishes running, the ability to get a runner registration code, etc).

Once we merge that, I will have a look at forgejo as well, but they seem to be going a different route in terms of runner management:

https://codeberg.org/forgejo/discussions/issues/241

**Reply by gabriel-samfira** [reply]
The PR landed in main. If you would like to test it out, there is a quick guide here: https://github.com/cloudbase/garm/blob/main/doc/gitea.md

**Reply by adamcstephens** [reply]
Thanks, that's amazing. Unfortunately I don't have a gitea instance to test against as I only use forgejo. Do you think it'll be possible to extend this to forgejo or are you still yet to investigate this?

**Reply by gabriel-samfira** [reply]
Not yet. Gitea has ephemeral runners and the workflow_job web hook which we need. Forgejo doesn't (yet) have ephemeral runners from what I see. They have the `--one-job` (or similar) flag for the runners, but if I remember correctly, the runner doesn't get de-registered.

So I need to investigate. The only other thing we need is some way to know when a job is available, either through a webhook or some other mechanism like longpoll (github scale sets for example), gRPC or websocket. If these 2 components are there, we can easily add forgejo as well.

I started with gitea because @ChristopherHX was kind enough to add the needed bits in act_runner and Gitea to enable GARM to easily integrate it. Forgejo is next, but I need to switch to something else for the next few weeks.

**Reply by DragonHunter274** [reply]
it looks like ephemeral runners have just landed in forgejo: https://code.forgejo.org/forgejo/runner/pulls/1122
just waiting for this now: https://codeberg.org/forgejo/forgejo/pulls/9803

**Reply by gabriel-samfira** [reply]
ohh wow!! That is awesome!!! I will definitely add this when all the bits land!

I was actually adding "agent mode" in an attempt to try to somehow consume the KEDA bits that were added a while ago, but the above bits would make things soooo much easier.

**Reply by zachjensz** [reply]
Now that Gitea support has landed and Forgejo supports ephemeral runners, is there a rough timeline for Forgejo support in GARM? Edit: I see this one is still a blocker https://codeberg.org/forgejo/forgejo/issues/9802
